Include/internal/pycore_parking_lot.h

```diff
@@ -0,0 +1,129 @@
+// ParkingLot is an internal API for building efficient synchronization
+// primitives like mutexes and events.
+//
+// The API and name is inspired by WebKit's WTF::ParkingLot, which in turn
+// is inspired Linux's futex API.
+// See https://webkit.org/blog/6161/locking-in-webkit/.
+//
+// The core functionality is an atomic "compare-and-sleep" operation along with
+// an atomic "wake-up" operation.
+
+#ifndef Py_INTERNAL_PARKING_LOT_H
+#define Py_INTERNAL_PARKING_LOT_H
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifndef Py_BUILD_CORE
+#  error "this header requires Py_BUILD_CORE define"
+#endif
+
+#include "pycore_time.h"          // _PyTime_t
+#include "cpython/pyatomic.h"
+
+
+enum {
+    // The thread was unparked by another thread.
+    Py_PARK_OK = 0,
+
+    // The value of `address` did not match `expected`.
+    Py_PARK_AGAIN = -1,
+
+    // The thread was unparked due to a timeout.
+    Py_PARK_TIMEOUT = -2,
+
+    // The thread was interrupted by a signal.
+    Py_PARK_INTR = -3,
+};
+
+// Checks that `*address == *expected` and puts the thread to sleep until an
+// unpark operation is called on the same `address`. Otherwise, the function
+// returns `Py_PARK_AGAIN`. The comparison is performed atomically
+// with respect to unpark operations.
+//
+// The `address_size` argument is the size of the data pointed to by the
+// `address` and `expected` pointers (i.e., sizeof(*address)).
+//
+// `arg`, which can be NULL, is passed to the unpark operation.
+//
+// The `timeout_ns` argument specifies the maximum amount of time to wait, with
+// -1 indicating an infinite wait.
+PyAPI_FUNC(int)
+_PyParkingLot_Park(const void *address, const void *expected,
+                   size_t address_size, _PyTime_t timeout_ns,
+                   void *arg, int detach);
+
+struct _PyUnpark {
+    // The `arg` value passed to _PyParkingLot_Park().
+    void *arg;
+
+    // Are there more threads waiting on the address? May be true in cases
+    // where threads are waiting on a different address that maps to the same
+    // internal bucket.
+    int more_waiters;
+};
+
+// Unpark a single thread waiting on `address`.
+//
+// The `unpark` is a pointer to a `struct _PyUnpark`.
+//
+// Usage:
+//  _PyParkingLot_Unpark(address, unpark, {
+//      if (unpark) {
+//          void *arg = unpark->arg;
+//          int more_waiters = unpark->more_waiters;
+//          ...
+//      }
+//  });
+#define _PyParkingLot_Unpark(address, unpark, ...)                          \
```

Comment:
This could be a function pointer. It may be a little more clunky because of the lack of closures in C, but it's doable. Once you get a chance to look at the usage (in _PyMutex_TryUnlock), let me know if you prefer the function pointer.